<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from tm5_900_robot.urdf.xacro       | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="tm5_900">
  <gazebo>
    <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so"> 
      <parameters>$(find my_robot_gazebo)/config/joint_controller.yaml</parameters>
     </plugin>  
    

    <!--
    <plugin name="gazebo_ros_power_monitor_controller" filename="libgazebo_ros_power_monitor.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1.0</updateRate>
      <timeout>5</timeout>
      <powerStateTopic>power_state</powerStateTopic>
      <powerStateRate>10.0</powerStateRate>
      <fullChargeCapacity>87.78</fullChargeCapacity>     
      <dischargeRate>-474</dischargeRate>
      <chargeRate>525</chargeRate>
      <dischargeVoltage>15.52</dischargeVoltage>
      <chargeVoltage>16.41</chargeVoltage>
    </plugin>
-->
  </gazebo>
  <!--xacro:include filename="$(find my_robot_gazebo)/urdf/tm.gazebo.xacro" /-->
  <!-- Kinematic model -->
  <!-- Properties from tmcontrol.conf -->
  <!--
      DH for TM5-900:
      a = [0.0000, 0.4290, 0.4115,  0.0000,  0.0000,  0.0000]
      d = [0.1452,  0.0000,  0.0000,  -0.1223,  0.1060,  0.11315]
      alpha = [-0.5*pi, 0, 0, 0.5*pi, 0.5*pi, 0]
      q_home_offset = [0, -0.5*pi, 0, 0.5*pi, 0, 0]
      joint_direction = [-1, -1, 1, -1, 1, 1]
      mass = [4.5, 10.5, 2.1, 1.45, 1.45, 0.4]
      center_of_mass = [ [0, -0.02561, 0.00193], [0.2125, 0, 0.11336], [0.11993, 0.0, 0.0265], [0, -0.0018, 0.01634], [0, 0.0018,0.01634], [0, 0, -0.001159] ](tm900)
    -->
  <!-- 0.1297 -->
  <!--xacro:property name="ee_length" value="${tm_d6 - wrist_3_length}" /-->
  <transmission name="shoulder_1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="shoulder_1_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="shoulder_1_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="shoulder_2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="shoulder_2_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="shoulder_2_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="elbow_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="elbow_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="elbow_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="wrist_1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wrist_1_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="wrist_1_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="wrist_2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wrist_2_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="wrist_2_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="wrist_3_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wrist_3_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="wrist_3_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <gazebo reference="shoulder_1_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_1_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="arm_2_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_1_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_2_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_3_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <!--
    <gazebo reference="${prefix}ee_link">
      <selfCollide>true</selfCollide>
    </gazebo>
    -->
  <link name="base_link">
    <visual>
      <geometry>
      
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm5_base.obj"/>
        
        <!--
         <mesh filename="/home/andrea/Scrivania/ros2_ws/src/my_robot_gazebo/meshes/tm5_900/visual/tm5_base.obj"/>
        -->
      </geometry>
      <material name="LightGrey">
        <color rgba="0.5 0.5 0.5 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm5_base_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.00110833289" ixy="0.0" ixz="0.0" iyy="0.0018" iyz="0.0" izz="0.00110833289"/>
    </inertial>
  </link>
  <joint name="shoulder_1_joint" type="revolute">
    <parent link="base_link"/>
    <child link="shoulder_1_link"/>
    <origin rpy="0 0 0" xyz="0 0.1452 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="150.0" lower="-4.71238898038469" upper="4.71238898038469" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="shoulder_1_link">
    <visual>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm_400w_01.obj"/>
      </geometry>
      <material name="DarkOlive">
        <color rgba="0.3 0.3 0.25 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm_400w_01_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="4.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.012487495005" ixy="0.0" ixz="0.0" iyy="0.0081" iyz="0.0" izz="0.012487495005"/>
    </inertial>
  </link>
  <joint name="shoulder_2_joint" type="revolute">
    <parent link="shoulder_1_link"/>
    <child link="arm_1_link"/>
    <origin rpy="0 0 0" xyz="0 0 0.146"/>
    <axis xyz="0 0 -1"/>
    <limit effort="150.0" lower="-3.141592653589793" upper="3.141592653589793" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="arm_1_link">
    <visual>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm5_900_arm1.obj"/>
      </geometry>
      <material name="DarkOlive">
        <color rgba="0.3 0.3 0.25 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm5_900_arm1_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="11.0"/>
      <origin rpy="0 0 0" xyz="0 0.2145 0"/>
      <inertia ixx="0.17860417855829996" ixy="0.0" ixz="0.0" iyy="0.019799999999999998" iyz="0.0" izz="0.17860417855829996"/>
    </inertial>
  </link>
  <joint name="elbow_joint" type="revolute">
    <parent link="arm_1_link"/>
    <child link="arm_2_link"/>
    <origin rpy="0 0 0" xyz="0 0.429 -0.12969999999999998"/>
    <axis xyz="0 0 -1"/>
    <limit effort="150.0" lower="-3.141592653589793" upper="3.141592653589793" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="arm_2_link">
    <visual>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm5_900_arm2.obj"/>
      </geometry>
      <material name="DarkOlive">
        <color rgba="0.3 0.3 0.25 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm5_900_arm2_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.5"/>
      <origin rpy="0 0 0" xyz="0 0.20575 0"/>
      <inertia ixx="0.0365431624660625" ixy="0.0" ixz="0.0" iyy="0.0025312499999999996" iyz="0.0" izz="0.0365431624660625"/>
    </inertial>
  </link>
  <joint name="wrist_1_joint" type="revolute">
    <parent link="arm_2_link"/>
    <child link="wrist_1_link"/>
    <origin rpy="0 0 0" xyz="0 0.4115 0.106"/>
    <axis xyz="0 0 -1"/>
    <limit effort="150.0" lower="-3.141592653589793" upper="3.141592653589793" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="wrist_1_link">
    <visual>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm_100w_01.obj"/>
      </geometry>
      <material name="DarkOlive">
        <color rgba="0.3 0.3 0.25 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm_100w_01_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.45"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.002474061510375" ixy="0.0" ixz="0.0" iyy="0.001468125" iyz="0.0" izz="0.002474061510375"/>
    </inertial>
  </link>
  <joint name="wrist_2_joint" type="revolute">
    <parent link="wrist_1_link"/>
    <child link="wrist_2_link"/>
    <origin rpy="0 0 0" xyz="0 0.106 0"/>
    <axis xyz="0 1 0"/>
    <limit effort="150.0" lower="-3.141592653589793" upper="3.141592653589793" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="wrist_2_link">
    <visual>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm_100w_02.obj"/>
      </geometry>
      <material name="DarkOlive">
        <color rgba="0.3 0.3 0.25 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm_100w_02_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.45"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.002474061510375" ixy="0.0" ixz="0.0" iyy="0.002474061510375" iyz="0.0" izz="0.001468125"/>
    </inertial>
  </link>
  <joint name="wrist_3_joint" type="revolute">
    <parent link="wrist_2_link"/>
    <child link="wrist_3_link"/>
    <origin rpy="0 0 0" xyz="0 0 0.11315"/>
    <axis xyz="0 0 1"/>
    <limit effort="150.0" lower="-4.71238898038469" upper="4.71238898038469" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="wrist_3_link">
    <visual>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/visual/tm_ee.obj"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.5 0.5 0.5 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://my_robot_gazebo/meshes/tm5_900/collision/tm_ee_c.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.4"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.000322499871" ixy="0.0" ixz="0.0" iyy="0.000322499871" iyz="0.0" izz="0.000405"/>
    </inertial>
  </link>
  <joint name="flange_fixed_joint" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="flange_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <link name="flange_link">
    </link>
  <!-- ROS base to TM 'base_link' Coordinates transform -->
  <link name="base"/>
  <joint name="base-base_link" type="fixed">
    <parent link="base"/>
    <child link="base_link"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
    <!-- Note: 0.5 * pi -->
  </joint>
  <!-- Frame coincident with all-zeros TCP on TM controller -->
  <link name="tool0"/>
  <joint name="flange_link-tool0" type="fixed">
    <parent link="flange_link"/>
    <child link="tool0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <link name="world"/>
  <joint name="world_joint" type="fixed">
    <parent link="world"/>
    <child link="base"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>


  <ros2_control name="GazeboSystem" type = "system">
    <hardware>
      <plugin>gabebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="wrist_2_joint">
      <command_interface name="position"> 
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="wrist_1_joint">
      <command_interface name="position"> 
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
  </ros2_control>
</robot>
